Added an SSE4 optimized adler32 checksum

KungFuJesus opened the pull request:
This variant uses the lower number of cycles psadw insruction in place
of pmaddubsw for the running sum that does not need multiplication.

This allows this sum to be done independently, partially overlapping the
running "sum2" half of the checksum.  We also have moved the shift
outside of the loop, breaking a small data dependency chain. The code
also now does a vectorized horizontal sum without having to rebase to
the adler32 base, as NMAX is defined as the maximum number of scalar
sums that can be peformed, so we're actually safe in doing this without
upgrading to higher precision.  We can do a partial horizontal sum
because psadw only ends up accumulating 16 bit words in 2 vector lanes,
the other two can safely be assumed as 0.

nmoinvaz commented on configure:
```diff
@@ -1063,7 +1063,30 @@ EOF
 
-check_sse4_intrinsics() {
+check_sse41_intrinsics() {
+    # Check whether compiler supports SSSE3 intrinsics
```
Should say SSE4.1 intrinsics.

nmoinvaz commented on configure:
```diff
@@ -1063,7 +1063,30 @@ EOF
+check_sse41_intrinsics() {
+    # Check whether compiler supports SSSE3 intrinsics
+    cat > $test.c << EOF
+#include <x86intrin.h>
```
Should be `#include <immintrin.h>`?

KungFuJesus replied:
Actually I think with sse4 it's supposed to be the pmmintrin, xmmintrin, etc specific includes.  Though, immintrin.h might work?

nmoinvaz commented on cmake/detect-intrinsics.cmake:
```diff
@@ -162,7 +162,35 @@ macro(check_ssse3_intrinsics)
+macro(check_sse42_intrinsics)
     if(CMAKE_C_COMPILER_ID MATCHES "Intel")
         if(CMAKE_HOST_UNIX OR APPLE)
             set(SSE4FLAG "-msse4.2")
```
Should be `SSE42FLAG`

nmoinvaz commented on arch/x86/adler32_sse41.c:
```diff
@@ -0,0 +1,103 @@
+#include <immintrin.h>
+
+static inline uint32_t partial_hsum(__m128i x)
+{
```
Bracket should be on previous line.

nmoinvaz commented on arch/x86/adler32_sse41.c:
```diff
@@ -0,0 +1,103 @@
+}
+
+static inline uint32_t hsum(__m128i x)
+{
```
Bracket should be on previous line.

nmoinvaz commented on configure:
```diff
@@ -102,7 +102,7 @@ forcesse2=0
 ssse3flag="-mssse3"
-sse4flag="-msse4"
+sse41flag="-msse4.1"
 sse42flag="-msse4.2"
```
Need to update these two lines in configure script:
```
echo sse4flag = $sse4flag >> configure.log
```
We should also add one for sse42flag.

And then update this one:
```
/^SSE4FLAG *=/s#=.*#=$sse4flag#
```
And add one for SSE42.

nmoinvaz commented on CMakeLists.txt:
```diff
@@ -724,24 +726,34 @@ if(WITH_OPTIM)
-            check_sse4_intrinsics()
+        if(WITH_SSE41)
+            check_sse41_intrinsics()
+            if(HAVE_SSE41_INTRIN)
```
Should have:
```
else() 
    set(WITH_SSE41 OFF) 
endif()
```
if `NOT HAVE_SSE41_INTRIN`

nmoinvaz commented on cmake/detect-intrinsics.cmake:
```diff
@@ -212,7 +212,35 @@ macro(check_ssse3_intrinsics)
+        endif()
+    endif()
+    # Check whether compiler supports SSSE3 intrinsics
+    set(CMAKE_REQUIRED_FLAGS "${SSE41FLAG}")
```
Should say SSE4.1

nmoinvaz commented on cmake/detect-intrinsics.cmake:
```diff
@@ -212,7 +212,35 @@ macro(check_ssse3_intrinsics)
+            set(SSE41FLAG "-msse4.1")
+        endif()
+    endif()
+    # Check whether compiler supports SSSE3 intrinsics
```
Should say SSE4.1

mtl1979 commented on arch/x86/adler32_sse41.c:
```diff
@@ -0,0 +1,101 @@
+       adler = partial_hsum(vs1) % BASE;
+       sum2 = hsum(vs2) % BASE;
+
+       vs1 = _mm_cvtsi32_si128(adler);
+       vs2 = _mm_insert_epi32(_mm_setzero_si128(), sum2, 3);
```
Why are these using different intrinsics even though both move 32-bit integer to vector register at same position.

KungFuJesus replied:
They don't.  The top one moves it at the first position (first 4 bytes in memory with little endian).  The other moves things to the last 4 bytes.  This is needed so that the multiplication done in the madd lines up.  You've been in big endian PPC land for too long :).  But also, there's a specific speed advantage to this - movd is cheap and it puts our sums in a lane that is non-zero, meaning we can get away with a partial horizontal sum instead of a full one.

KungFuJesus replied:
> README.md also needs to be updated in two places.

will do

mtl1979 replied:
As this differs from other adler32 versions, it should be properly commented...

KungFuJesus replied:
Hmm, don't know why github's quote reply somehow got attached to this thread...

mtl1979 replied:
I've always thought `_mm_insert_epi*` instructions are expensive instructions, basically taking 2 cycles, combined with 0.33 cycles for _mm_setzero_si128(), which also requires own register as it's basically implemented as xor with itself. Both `_mm_cvtsi32_si128` and shifting the vector contents take 1 cycles, which should save 0.33 cycles and one vector register.

KungFuJesus replied:
Well there's a few things happening here. The xor to set to zero is cheap and we can do a lot of them at once. But additionally, it's a zeroing idiom which may in fact make the register rename engine in the register file just quickly swap in a zero'd register. In addition to that, I don't think that the movd instruction is one cycle. According to agner fog's instruction table, on wolfdale movd between GPR and XMM register is 2 cycles.

KungFuJesus replied:
Also the shifting of bytes still requires the load to complete. The best we could hope for is that it'd overlap the movd that precedes it. But that shift puts us outside of that 2 cycle range for overlap.

KungFuJesus replied:
And also the same seems to be true for Nehalem as well:
https://www.agner.org/optimize/instruction_tables.pdf
Though, he qualifies the latency a little better in those tables.  The latency is 1+1, where there's a bypass delay that has to occur when crossing the GPR/XMM boundary.

KungFuJesus replied:
> As this differs from other adler32 versions, it should be properly commented...

I actually don't know that it necessarily needs to differ from the other implementations.  The front half of the checksum never actually required the existing running sum to be in any particular position in the vector register.  It was only the second half (sum2) that needed this.  We could just as easily modify the ssse3 variant in this commit to do the same exact thing.  In fact, I've already done so on my avx512 PR.

mtl1979 replied:
I don't really trust any tables listing instruction latencies and throughputs as the real-life values highly depend on how the processor and compiler reorder the instructions to utilize as many concurrent ports as possible.

In both cases it has to move data from GPR to xmm register, which means there is delay when it crosses from one execution unit to another. Usually the microcode is optimized better when moving to vector element 0, otherwise it wouldn't make sense to have separate instruction mnemonic for that.

It would be ideal if we could avoid trip to GPR register during the horizontal sum, but that is unlikely to be possible with any of the current processors as modulo is really expensive operation using vector registers.

KungFuJesus replied:
Generally speaking these values are found with some pretty careful experimentation, but I think Intel's documentation will probably confirm the same.
In any case, even though on newer architectures (as in, ivy bridge and newer), it's only 1 cycle to do the GPR -> XMM move (as documented via Intel's own intrinsic guide, anyway), you still have to wait for the move to complete before doing the byte shift, meaning this would take net 2 cycles. 

Now where this might make a difference is if you're dealing with ymm registers, as _those_ generally need to expand to a pinsert instruction + a combination of 2 XMM registers.  For that, it makes way more sense to do a broadcast + and with a static mask ymm register, as that is likely to be the cheapest way to do it.

KungFuJesus replied:
Hah, actually...it's not even important for us to put sum2 in the last vector position.  I'm not exactly sure why any of this code was doing this, as it just needs to be summed, not multiplied at all.  Interesting...this certainly resolves this discussion, at least.

KungFuJesus replied:
Well that's rich...I modified the AVX512 variant to use the two simpler vmovd instructions to xmm registers, allowing the register aliasing to zero the other bits when it turns into a zmm register and...it was slower.  I think there's a penalty when doing that, but the weird part is we should only be seeing at first register access which would be at the top of the loop and every time the loop did horizontal sums to prevent overflow (right at NMAX).  Either that, or we're able to do things on more execution ports somehow by mixing the vbroadcastd (with mask) with the vmovd instruction.  

It's weird because the compiler does this trick all the time with xmm registers.  It often will pxor an xmm register for zero before it gets used as a ymm register.  I guess I just leave the avx512 version as-is for now, populating sum2 in the last slot for now (using the first slot turns on an optimization that makes the compiler use vmovd instead).

KungFuJesus replied:
So, it definitely doesn't seem to be an instruction alignment thing.  Inserting two NOPs after the vmovd's to give the same lengths (while keeping the bottom of the loop where vs1 and vs2 are reset) doesn't seem to do anything for speed on the avx512 vnni version.  We must be hitting some sort of transition penalty from xmm->zmm when both vs1 and vs2 are set as xmm registers. Even weirder, that's what the vex/evex encoded instructions are supposed to prevent.

nmoinvaz commented:
README.md also needs to be updated in two places.

KungFuJesus commented:
Here's the benchmark results on this ancient Q9650:
```
mythtv@mythtvbox ~/zlib-ng/build2 $ ./adler32_benchmark --benchmark_repetitions=10
2021-12-31T14:53:22-05:00
Running ./adler32_benchmark
Run on (4 X 2997.22 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 6144 KiB (x2)
Load Average: 0.29, 0.12, 0.10
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
adler32_c_bench                 69924 ns        69924 ns        70226
adler32_c_bench                 69877 ns        69877 ns        70226
adler32_c_bench                 69942 ns        69940 ns        70226
adler32_c_bench                 69997 ns        69996 ns        70226
adler32_c_bench                 70007 ns        70007 ns        70226
adler32_c_bench                 69981 ns        69978 ns        70226
adler32_c_bench                 69925 ns        69923 ns        70226
adler32_c_bench                 69868 ns        69868 ns        70226
adler32_c_bench                 69873 ns        69872 ns        70226
adler32_c_bench                 69879 ns        69878 ns        70226
adler32_c_bench_mean            69927 ns        69926 ns           10
adler32_c_bench_median          69925 ns        69924 ns           10
adler32_c_bench_stddev           53.3 ns         53.0 ns           10
adler32_c_bench_cv               0.08 %          0.08 %            10
adler32_sse41_bench             31227 ns        31227 ns       100000
adler32_sse41_bench             31259 ns        31259 ns       100000
adler32_sse41_bench             31279 ns        31278 ns       100000
adler32_sse41_bench             31250 ns        31248 ns       100000
adler32_sse41_bench             31265 ns        31263 ns       100000
adler32_sse41_bench             31257 ns        31256 ns       100000
adler32_sse41_bench             31264 ns        31264 ns       100000
adler32_sse41_bench             31286 ns        31285 ns       100000
adler32_sse41_bench             31276 ns        31275 ns       100000
adler32_sse41_bench             31275 ns        31274 ns       100000
adler32_sse41_bench_mean        31264 ns        31263 ns           10
adler32_sse41_bench_median      31265 ns        31264 ns           10
adler32_sse41_bench_stddev       16.9 ns         17.1 ns           10
adler32_sse41_bench_cv           0.05 %          0.05 %            10
adler32_ssse3_bench             34701 ns        34700 ns       100000
adler32_ssse3_bench             34698 ns        34698 ns       100000
adler32_ssse3_bench             34698 ns        34698 ns       100000
adler32_ssse3_bench             34689 ns        34688 ns       100000
adler32_ssse3_bench             34693 ns        34693 ns       100000
adler32_ssse3_bench             34703 ns        34703 ns       100000
adler32_ssse3_bench             34706 ns        34705 ns       100000
adler32_ssse3_bench             34696 ns        34693 ns       100000
adler32_ssse3_bench             34694 ns        34692 ns       100000
adler32_ssse3_bench             34696 ns        34696 ns       100000
adler32_ssse3_bench_mean        34697 ns        34697 ns           10
adler32_ssse3_bench_median      34697 ns        34697 ns           10
adler32_ssse3_bench_stddev       4.87 ns         5.40 ns           10
adler32_ssse3_bench_cv           0.01 %          0.02 %            10
```
It's about a microsecond faster than this when compiled with -mtune=native, of course, but I guess it's more fair to go based on generic flags.  I suspect sandybridge, the latest architecture that could benefit from this because it lacks AVX2, might even have a more optimal instruction scheduling sequence with native compiler flags.  Unfortunately my cascadelake system replaced my sandybridge-E one.

KungFuJesus commented:
Oh man, aligned made a pretty substantial difference on Yorkfield, still:
```
mythtv@mythtvbox ~/zlib-ng/build2 $ ./adler32_benchmark 
2021-12-31T17:49:36-05:00
Running ./adler32_benchmark
Run on (4 X 2997.22 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 6144 KiB (x2)
Load Average: 0.28, 0.13, 0.04
--------------------------------------------------------------
Benchmark                    Time             CPU   Iterations
--------------------------------------------------------------
adler32_c_bench          70057 ns        70056 ns        69203
adler32_ssse3_bench      34470 ns        34469 ns       100000
adler32_sse41_bench      26094 ns        26094 ns       100000
```
Perhaps it _is_ worth writing a version that eats the initial load as unaligned and then does subsequent aligned loads.  On Nehalem and later, likely not _as_ big of an impact, but wow.  Worth noting of course that using the unaligned instructions with aligned memory did not realize these benefits like they ought to with Nehalem and newer.

mtl1979 commented on arch/x86/Makefile.in:
```diff
@@ -11,7 +11,8 @@ SUFFIX=
 SSSE3FLAG=-mssse3
-SSE4FLAG=-msse4
+SSE41FLAG=-msse4.1
+SSE42FLAG=-msse4.2 -mcrc32
```
Clang doesn't like `-mcrc32`.

KungFuJesus replied:
Hmm I stumbled onto gcc failing to build this due to use native being flipped in. Still, likely gcc won't struggle with this with -msse4.2, even though it's less specific. I'll change it back.

KungFuJesus commented:
Very _nearly_ have a palignr leveraging version of this, but it's not quite correct yet.  That should be faster on most every CPU pre Nehalem (and shouldn't be a meaningful penalty post-Nehalem).  I may have to pick this up tomorrow, though.

Edit: ugh, it's kind of tricky to get this right.  Using palignr with a giant jump table is feasible and not that bad to generate with macros, but the real issue becomes the shifting used to multiply by 16 for the sum2 portion of the sum.  We can only ever do this after dealing with a full set of 16 values, and it's hard to do the remainder peeling such that we don't end up loading beyond a multiple of 16 so that this multiplication is allowed.  It's a shame that the natural alignment for an array of characters happens to usually be 1 byte.  We could perhaps write an aligned/vs unaligned label to jump to for when we get lucky, but it's ultimately up to the caller to decide where the memory came from for the adler32 input.  _Maybe_ some gains could be had where we made zlib-ng do at least 16 byte aligned allocations for streams generated for deflate and get some benefits there.  

My use case for zlib-ng is namely accelerate libpng, which absolutely crawls with vanilla zlib.  I somewhat suspect that allocation will be done with libpng rather than zlib directly for the streaming input so that's perhaps a little less than ideal (then again, I've never actually checked the backing allocator for things fed from libpng, maybe it does use zlib for what get's adler32 checksummed?).  

Does anyone have any idea if this is in the realm of possibilities?  At least on *nix platforms, where posix_memalign / aligned_alloc are possible?  Both of those allocators at least are capable of being freed by the standard free, which is great.  It'd be trivial to write an alternate loop that takes advantage of this alignment.

nmoinvaz commented:
>  I somewhat suspect that allocation will be done with libpng rather than zlib directly for the streaming input...

It looks like libpng does use its own allocation functions.
https://github.com/glennrp/libpng/blob/a37d4836519517bdce6cb9d956092321eca3e73b/png.c#L96

nmoinvaz commented:
> Using palignr with a giant jump table is feasible..

I don't think it would be ideal if it increases code size with large table. We had a large table at one time in deflate_quick and it ended up being faster, but not much faster to warrant the extra size.

mtl1979 commented:
@nmoinvaz libpng's default allocation function is just wrapper around malloc, which will not cause issues in most cases as malloc already aligns memory so most vector instructions are happy.

KungFuJesus commented:
> @nmoinvaz libpng's default allocation function is just wrapper around malloc, which will not cause issues in most cases as malloc already aligns memory so most vector instructions are happy.

I'm pretty sure malloc by itself gives no natural alignment. It looks like zlib has a malloc wrapper (zng_alloc) that actually attempts to align things to make vector instructions happy, but from what I can see from deflate, there's a fair bit of arbitrary seeking going on that we can't reliably have aligned access every time. Still, opportunistically doing an aligned load version of the loop doesn't seem like it'd be an awful compromise.

As far as the jump table, it's away from the main code and it's only entered once to realign the multiplication operand and shift a mask so that the product ends up being the same. I still don't have it quite right though, there's probably a detail I'm missing that's important for the intermediate sums in vs2.  For now it's stashed but just having an aligned variant of the loop may be the best way to go that doesn't complicate the code too badly.

mtl1979 commented:
@KungFuJesus On 64-bit platforms using malloc is safe in general as it aligns to 64-bit boundary or better. On 32-bit platforms it only guarantees 32-bit boundary, which can can cause issues if the code does 64-bit load directly instead of two separate 32-bit loads and then combining them.

KungFuJesus commented:
> @KungFuJesus On 64-bit platforms using malloc is safe in general as it aligns to 64-bit boundary or better. On 32-bit platforms it only guarantees 32-bit boundary, which can can cause issues if the code does 64-bit load directly instead of two separate 32-bit loads and then combining them.

I'll give you that, you'll probably get a natural 64 bit alignment to the pointer but in this case I'd need 128 bit (16 byte) alignment to safely satisfied an aligned vectorized load.

mtl1979 commented:
@KungFuJesus I haven't seen many allocating functions to guarantee 128-, 256- or 512-bit boundary but there is some already. In most cases too strict default alignment wastes a lot of memory as both start and end of allocation must be aligned to same boundary.

Inside the custom allocator the original pointer is usually stored in the previous 32 or 64 bits before the returned "user" pointer, so the inner pointer might need to be misaligned or otherwise there will be gap of unused bytes.

KungFuJesus commented:
> @KungFuJesus I haven't seen many allocating functions to guarantee 128-, 256- or 512-bit boundary but there is some already. In most cases too strict default alignment wastes a lot of memory as both start and end of allocation must be aligned to same boundary.
> 
> Inside the custom allocator the original pointer is usually stored in the previous 32 or 64 bits before the returned "user" pointer, so the inner pointer might need to be misaligned or otherwise there will be gap of unused bytes.

Yes though on platforms that support posix_memalign or aligned_alloc, the os actually gives you power of 2 alignment with no waste. For some reason Microsoft didn't deem it an important part of the c11 spec 🙄

mtl1979 commented:
@KungFuJesus As someone who has and still work with Microsoft, I can say none of the current mainstream operating systems follow the standards strictly. GNU style operating systems have own extensions for better backwards compatibility but Microsoft generally adds features only when needed and usually does not reuse existing code or file layout as all supported platforms use unified SDK.

Dead2 commented:
I'd like to see this split up into two commits (or PRs even), one for making the SSE4 rename changes that are needed in preparation for the optimized adler32 version. Makes it easier to review and test them separately, as well as potentially bisecting in the future.

KungFuJesus commented:
> I'd like to see this split up into two commits (or PRs even), one for making the SSE4 rename changes that are needed in preparation for the optimized adler32 version. Makes it easier to review and test them separately, as well as potentially bisecting in the future.

Will do.  I had a WIP with trying to get things functioning with aligned memory but it's proving kind of hard thanks to the requirement of carrying the previous sum of 16 elements from i-1's iteration.  Surprisingly, the other stuff was relatively easy to do with palignr with a fairly simple switch statement but I'm still a little hung up on how to efficiently carry through the prior sum in a way that works with the misalignment.  For now I think having an opportunistic version of the loop that uses aligned loads when we have an aligned buffer is the best we can do.  It's a little bit more code for basically no gain on everything after Nehalem but it's meaningful on everything that precedes it.

KungFuJesus committed:
Add SSE4.1 detection

Code leveraging this for the adler checksum is forthcoming

KungFuJesus committed:
Added an SSE4 optimized adler32 checksum

This variant uses the lower number of cycles psadw insruction in place
of pmaddubsw for the running sum that does not need multiplication.

This allows this sum to be done independently, partially overlapping the
running "sum2" half of the checksum.  We also have moved the shift
outside of the loop, breaking a small data dependency chain. The code
also now does a vectorized horizontal sum without having to rebase to
the adler32 base, as NMAX is defined as the maximum number of scalar
sums that can be peformed, so we're actually safe in doing this without
upgrading to higher precision.  We can do a partial horizontal sum
because psadw only ends up accumulating 16 bit words in 2 vector lanes,
the other two can safely be assumed as 0.

KungFuJesus commented:
@Dead2 This should be good to go.

KungFuJesus commented:
Hmm, this is...odd: https://github.com/zlib-ng/zlib-ng/pull/1094/conflicts
That hardly seems like a conflict, what's the issue there?

Edit: Oh I see, it's the latest commit on develop.

KungFuJesus commented:
Last minute testing may have found something in the CPU detection bits, hold off the merge for a bit, I'll see what's going on in the cpuid stuff.

Figured it out - I had a different install prefix.  Everything was fine :). @Dead2 this is good to go.

Dead2 reviewed (approved):
LGTM, but needs a rebase

KungFuJesus commented:
> LGTM, but needs a rebase

Just rebased

Dead2 commented:
This needs another rebase now 😉

KungFuJesus commented:
> This needs another rebase now wink

Done
